src/Compiler/Facilities/DiagnosticsLogger.fs

```diff
@@ -375,19 +375,29 @@ type CapturingDiagnosticsLogger(nm, ?eagerFormat) =
         let errors = diagnostics.ToArray()
         errors |> Array.iter diagnosticsLogger.DiagnosticSink
 
+
+type CompilationScopeHolder() =
+    member val buildPhase = BuildPhase.DefaultPhase with get, set
+    member val diagnosticsLogger = AssertFalseDiagnosticsLogger with get, set
+
 /// Type holds thread-static globals for use by the compiler.
 type internal DiagnosticsThreadStatics =
-    static let buildPhase = new AsyncLocal<BuildPhase option>()
+    static let holder = new AsyncLocal<CompilationScopeHolder option>()
```

Comment:
What are the cases when holder itself is not initialized? Maybe `Holder` method itself can be lazy instead of option?

Those are nitpicks anyway, just thinking out loud, so we don't have double initialization situations.

Reply:
Unilke ThreadLocal, there is no way to pass a factory func to AsyncLocal. It always has default value in new context.
So we need to explicitly check and initialize it somehow because we don't know the context. We will sort it out.